[workflow] Add limitations section (#3201)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/workflow/workflow-features-concepts.md

### Workflow replay

Dapr Workflows maintain their execution state by using a technique known as [event sourcing](https://learn.microsoft.com/azure/architecture/patterns/event-sourcing). Instead of storing the current state of a workflow as a snapshot, the workflow engine manages an append-only log of history events that describe the various steps that a workflow has taken. When using the workflow SDK, these history events are stored automatically whenever the workflow "awaits" for the result of a scheduled task.

When a workflow "awaits" a scheduled task, it unloads itself from memory until the task completes. Once the task completes, the workflow engine schedules the workflow function to run again. This second workflow function execution is known as a _replay_. 

> When a workflow function is replayed, it runs again from the beginning. However, when it encounters a task that it already scheduled, instead of scheduling that task again, the workflow engine:

Reviewer: I'm just realizing that it's confusing to talk about this task as being "scheduled". I think we should be using "completed" here instead.
```suggestion
When a workflow function is replayed, it runs again from the beginning. However, when it encounters a task that already completed, instead of scheduling that task again, the workflow engine:
```

> 1. Returns the result of the scheduled task to the workflow

Reviewer: ```suggestion
1. Returns the result of the completed task to the workflow
```

Using this replay technique, a workflow is able to resume execution from any "await" point as if it had never been unloaded from memory. Even the values of local variables from previous runs can be restored without the workflow engine knowing anything about what data they stored. This ability to restore state makes Dapr Workflows _durable_ and _fault tolerant_.

{{% alert title="Note" color="primary" %}}
The workflow replay behavior described here requires that workflow function code be _deterministic_. Deterministic workflow functions take the exact same actions when provided the exact same inputs. [Learn more about the limitations around deterministic workflow code.]({{< ref "workflow-features-concepts.md#workflow-determinism-and-code-constraints" >}})
{{% /alert %}}

### Infinite loops and eternal workflows

As discussed in the [workflow replay]({{< ref "#workflow-replay" >}}) section, workflows maintain a write-only event-sourced history log of all its operations. To avoid runaway resource usage, workflows must limit the number of operations they schedule. For example, ensure your workflow doesn't:

- Use infinite loops in its implementation

> - Schedule millions of tasks.

Reviewer: I know I'm commenting on the content that I wrote originally, but I'm thinking we should be more conservative here and say "thousands" of tasks instead of "millions". It's also more consistent with the guidance we have further below.

You can use the following two techniques to write workflows that may need to schedule extreme numbers of tasks:

1. **Use the _continue-as-new_ API**:  
    Each workflow SDK exposes a _continue-as-new_ API that workflows can invoke to restart themselves with a new input and history. The _continue-as-new_ API is especially ideal for implementing "eternal workflows", like monitoring agents, which would otherwise be implemented using a `while (true)`-like construct. Using _continue-as-new_ is a great way to keep the workflow history size small.

1. **Use child workflows**:  

>     Each workflow SDK exposes an API for creating child workflows. A child workflow mimics any other workflow, except that it's scheduled by a parent workflow. Child workflows have:

Reviewer: The word "mimic" seemed a little strange to me here. Could we maybe use "behaves like" instead?
```suggestion
    Each workflow SDK exposes an API for creating child workflows. A child workflow behaves like any other workflow, except that it's scheduled by a parent workflow. Child workflows have:
```

1. **Workflow functions must call deterministic APIs.**  
    APIs that generate random numbers, random UUIDs, or the current date are _non-deterministic_. To work around this limitation, you can:
    - Use these APIs in activity functions, or 
    - (Preferred) Use built-in equivalent APIs offered by the SDK. For example, each authoring SDK provides an API for retrieving the current time in a deterministic manner.  

    For example, instead of this:

>     ```csharp

Reviewer: Should we set up language tabs here since we'll want to add examples for multiple languages? Or would you prefer that we add language tabs later, when we actually have multiple languages?

Author: good idea - added a tab to prep for languages being added!
